src/servers/udp/server.rs

```diff
@@ -252,124 +329,178 @@ impl Udp {
+        tracing::debug!(target: "UDP TRACKER: Udp::run_with_graceful_shutdown", local_addr, "(started)");
 
         let stop = running.abort_handle();
 
         select! {
-            _ = running => { debug!(target: "UDP TRACKER", "Socket listener stopped on address: udp://{address}"); },
-            () = halt => { debug!(target: "UDP TRACKER", "Halt signal spawned task stopped on address: udp://{address}"); }
+            _ = running => { tracing::debug!(target: "UDP TRACKER: Udp::run_with_graceful_shutdown", local_addr, "(stopped)"); },
+            _ = halt_task => { tracing::debug!(target: "UDP TRACKER: Udp::run_with_graceful_shutdown",local_addr, "(halting)"); }
         }
         stop.abort();
 
-        task::yield_now().await; // lets allow the other threads to complete.
+        tokio::task::yield_now().await; // lets allow the other threads to complete.
     }
 
-    async fn run_udp_server(tracker: Arc<Tracker>, socket: Arc<UdpSocket>) {
-        let tracker = tracker.clone();
-        let socket = socket.clone();
+    async fn run_udp_server_main(mut direct: Receiver) {
+        let reqs = &mut ActiveRequests::default();
 
-        let reqs = &mut ActiveRequests::new().await;
+        let addr = direct.socket.local_addr().expect("it should get local address");
+        let local_addr = format!("udp://{addr}");
 
         loop {
-            task::yield_now().await;
-            for h in reqs.rb.iter_mut() {
-                if h.is_finished() {
-                    std::mem::swap(
-                        h,
-                        &mut Self::spawn_request_processor(
-                            Self::receive_request(socket.clone()).await,
-                            tracker.clone(),
-                            socket.clone(),
-                        )
-                        .abort_handle(),
-                    );
-                } else {
-                    // the task is still running, lets yield and give it a chance to flush.
+            if let Some(req) = {
+                tracing::trace!(target: "UDP TRACKER: Udp::run_udp_server", local_addr, "(wait for request)");
+                direct.next().await
+            } {
+                tracing::trace!(target: "UDP TRACKER: Udp::run_udp_server::loop", local_addr, "(in)");
+
+                let req = match req {
+                    Ok(req) => req,
+                    Err(e) => {
+                        if e.kind() == std::io::ErrorKind::Interrupted {
+                            tracing::warn!(target: "UDP TRACKER: Udp::run_udp_server::loop", local_addr, err = %e,  "(interrupted)");
+                            return;
+                        }
+                        tracing::error!(target: "UDP TRACKER: Udp::run_udp_server::loop", local_addr, err = %e,  "break: (got error)");
+                        break;
+                    }
+                };
+
+                if req.is_finished() {
+                    continue;
+                }
+
```

Comment:
Hi @da2ce7 I guess from this point on, we have created a new task that has not been finished yet and we need to add it to the "pending to finish" limited buffer. Why don't we extract a type "buffer" for that?

**ChatGPT refactor (not tested).**

To make the main loop clearer, you can extract the request handling logic into a new `RequestBuffer` type. Here's how you can refactor the code to achieve this:

1. **Define `RequestBuffer` Type**:
   Define a struct to encapsulate the buffer logic for pending requests.

2. **Refactor the Main Loop**:
   Move the buffer handling logic from the main loop into methods on the `RequestBuffer`.

Here’s a possible implementation of this refactor:

### Step 1: Define the `RequestBuffer` Type

```rust
use std::collections::VecDeque;

struct RequestBuffer<T> {
    buffer: VecDeque<T>,
    max_size: usize,
}

impl<T> RequestBuffer<T> {
    fn new(max_size: usize) -> Self {
        Self {
            buffer: VecDeque::with_capacity(max_size),
            max_size,
        }
    }

    fn try_push(&mut self, item: T) -> Result<(), T> {
        if self.buffer.len() < self.max_size {
            self.buffer.push_back(item);
            Ok(())
        } else {
            Err(item)
        }
    }

    fn pop_iter(&mut self) -> impl Iterator<Item = T> + '_ {
        self.buffer.drain(..)
    }
}
```

### Step 2: Refactor the Main Loop

Replace the buffer handling code in `run_udp_server_main` with calls to `RequestBuffer` methods:

```rust
impl Udp {
    async fn run_udp_server_main(mut direct: Receiver) {
        let reqs = &mut ActiveRequests::default();
        let addr = direct.socket.local_addr().expect("it should get local address");
        let local_addr = format!("udp://{addr}");

        // Create a RequestBuffer instance with the desired max size
        let mut request_buffer = RequestBuffer::new(MAX_BUFFER_SIZE);

        loop {
            if let Some(req) = {
                tracing::trace!(target: "UDP TRACKER: Udp::run_udp_server", local_addr, "(wait for request)");
                direct.next().await
            } {
                tracing::trace!(target: "UDP TRACKER: Udp::run_udp_server::loop", local_addr, "(in)");

                let req = match req {
                    Ok(req) => req,
                    Err(e) => {
                        if e.kind() == std::io::ErrorKind::Interrupted {
                            tracing::warn!(target: "UDP TRACKER: Udp::run_udp_server::loop", local_addr, err = %e,  "(interrupted)");
                            return;
                        }
                        tracing::error!(target: "UDP TRACKER: Udp::run_udp_server::loop", local_addr, err = %e,  "break: (got error)");
                        break;
                    }
                };

                if req.is_finished() {
                    continue;
                }

                // Try to push the request into the buffer
                if let Err(req) = request_buffer.try_push(req) {
                    // If the buffer is full, process pending requests
                    let mut finished: u64 = 0;
                    let mut unfinished_task = None;

                    for h in request_buffer.pop_iter() {
                        if h.is_finished() {
                            finished += 1;
                            continue;
                        }

                        tokio::task::yield_now().await;

                        if h.is_finished() {
                            finished += 1;
                            continue;
                        }

                        tracing::debug!(target: "UDP TRACKER: Udp::run_udp_server::loop",  local_addr, removed_count = finished, "(got unfinished task)");

                        if finished == 0 {
                            h.abort();
                            tracing::warn!(target: "UDP TRACKER: Udp::run_udp_server::loop",  local_addr, "aborting request: (no finished tasks)");
                            break;
                        }

                        unfinished_task = Some(h);
                    }

                    if let Some(h) = unfinished_task {
                        request_buffer.try_push(h).expect("it was previously inserted");
                    }

                    if !req.is_finished() {
                        request_buffer.try_push(req).expect("it should remove at least one element.");
                    }
                }
            } else {
                tokio::task::yield_now().await;
                tracing::error!(target: "UDP TRACKER: Udp::run_udp_server",  local_addr, "breaking: (ran dry, should not happen in production!)");
                break;
            }
        }
    }

    // Other methods...
}
```

### Summary

1. **RequestBuffer Definition**:
   The `RequestBuffer` struct manages a buffer of pending requests, encapsulating the logic for adding and removing requests.

2. **Main Loop Refactoring**:
   The main loop in `run_udp_server_main` is simplified by moving buffer management code into `RequestBuffer`.

This refactoring makes the main loop cleaner and more focused on handling UDP requests, while the `RequestBuffer` type manages the state and logic related to buffering and processing pending requests.

Reply:
Merged in: https://github.com/torrust/torrust-tracker/pull/913